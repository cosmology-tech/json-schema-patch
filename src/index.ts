import { applyPatch, Operation } from 'fast-json-patch';
import { BooleanFunction, createJSONSchemaPatchOperations, dirname, findAllProps, TransformFunction } from './utils';

export interface JSONSchemaPatchOperation {
  op: 'add' | 'remove' | 'replace' | 'rename' | 'addProperty' | 'removeProperty' | 'renameProperty' | 'addDefinition' | 'removeDefinition';
  path: string;
  value?: any;
}

function getValueAtPath(obj, path) {
  const parts = path.split('/').slice(1); // Split the path and remove the initial empty string
  let current = obj;
  for (let part of parts) {
    if (current[part] === undefined) return undefined; // Path does not exist
    current = current[part];
  }
  return current;
}

export function expandOperations(operations: JSONSchemaPatchOperation[]): JSONSchemaPatchOperation[] {
  const expandedOps: JSONSchemaPatchOperation[] = [];

  operations.forEach(op => {
    if (Array.isArray(op.value)) {
      // Create a new operation for each item in the value array
      op.value.forEach(item => {
        expandedOps.push({
          op: op.op,
          path: op.path,
          value: item
        });
      });
    } else if (typeof op.value !== 'undefined') {
      // Directly push the operation if the value is defined
      expandedOps.push(op);
    } else {
      // Push the operation if there's no value, such as for 'remove' operations
      expandedOps.push({
        op: op.op,
        path: op.path
      });
    }
  });

  return expandedOps;
}

function sanitizeJsonPath(path: string): string {
  // Remove redundant slashes and ensure there is no trailing slash.
  const sanitizedPath = path.replace(/\/+/g, '/').replace(/\/$/, '');
  // Ensure the path starts with a slash unless it's an empty string.
  return sanitizedPath.startsWith('/') || sanitizedPath === '' ? sanitizedPath : '/' + sanitizedPath;
}

export class JSONSchemaPatch {
  private ops: JSONSchemaPatchOperation[] = [];

  constructor(public schema: any) { }

  // Accumulate operations in an internal state
  prepareOperation(op: JSONSchemaPatchOperation) {
    this.ops.push(op);
  }

  transform(transformFunction: TransformFunction, transformTest: BooleanFunction): void {
    this.ops = [
      ...this.ops,
      ...createJSONSchemaPatchOperations(findAllProps(this.schema), transformFunction, transformTest)
    ];
  }

  // Apply all accumulated operations
  applyPatch(): any {
    const ops = expandOperations(this.ops);
    ops.forEach(op => {
      switch (op.op) {
        case 'addProperty':
          this.addProperty(op.path, op.value.name, op.value.property);
          break;
        case 'removeProperty':
          this.removeProperty(op.path, op.value);
          break;
        case 'renameProperty':
          this.renameProperty(op.path, op.value.oldName, op.value.newName);
          break;
        case 'rename':
          this.rename(op.path, op.value);
          break;
        case 'replace':
          this.replace(op.path, op.value);
          break;
        case 'add':
          this.add(op.path, op.value);
          break;
        case 'remove':
          this.remove(op.path);
          break;
        case 'addDefinition':
          this.addDefinition(op.path, op.value);
          break;
        case 'removeDefinition':
          this.removeDefinition(op.value);
          break;
        default:
          throw new Error('OP not supported!');
      }
    });

    // Reset the operations array after application
    this.ops = [];
    return this.schema;
  }

  // Add a definition to the $defs section
  addDefinition(name: string, definition: any): void {
    const path = sanitizeJsonPath(`/$defs/${name}`);
    // Create an operation to add the new definition
    const op: Operation = { op: 'add', path, value: definition };
    // Apply the operation using fast-json-patch
    console.log(op)
    applyPatch(this.schema, [op]);
  }

  // Remove a definition from the $defs section and any references to it
  removeDefinition(name: string): void {
    // Construct the path to the definition
    const path = sanitizeJsonPath(`/$defs/${name}`);
    // First, remove any properties that reference this definition
    this.removePropertiesUsingDefinition(name);
    // Then, remove the definition itself
    const op: Operation = { op: 'remove', path };
    applyPatch(this.schema, [op]);
  }

  // Helper method to remove properties that reference a specific definition
  private removePropertiesUsingDefinition(defName: string): void {
    const findPaths = (obj: any, currentPath: string) => {
      Object.entries(obj).forEach(([key, value]) => {
        if (typeof value === 'object' && value !== null) {
          // Check if the object has a $ref that matches the definition
          //@ts-ignore
          if (value.$ref && value.$ref === `#/$defs/${defName}`) {
            // Determine if we're within a properties object
            if (currentPath.endsWith('/properties')) {
              // Remove property by calling removeProperty
              // Extract the parent path and the property name
              let propertyPath = currentPath.replace('/properties', '');
              this.removeProperty(propertyPath, key);
            } else {
              // If it's not a direct property, just remove the reference
              const fullPath = currentPath + '/' + key;
              const op: Operation = { op: 'remove', path: fullPath };
              applyPatch(this.schema, [op]);
            }
          }
          // Recurse further into the object
          findPaths(value, currentPath + '/' + key);
        }
      });
    };

    // Start the recursive search from the root of the schema
    findPaths(this.schema, '');
  }

  // Add a property at a specified path
  addProperty(path: string, propertyName: string, property: any): void {
    path = path.trim() === '/' ? '' : path;
    const fullPath = sanitizeJsonPath(`${path}/properties/${propertyName}`);
    const op: Operation = { op: 'add', path: fullPath, value: property };
    applyPatch(this.schema, [op]);
  }

  // Remove a property from a specified path
  removeProperty(path: string, propertyName: string): void {
    path = path.trim() === '/' ? '' : path;
    const fullPath = sanitizeJsonPath(`${path}/properties/${propertyName}`);
    const op: Operation = { op: 'remove', path: fullPath };
    applyPatch(this.schema, [op]);
    this.removeRequiredFromProperty(path, propertyName);
  }

  // Remove from required if the property is present
  private removeRequiredFromProperty(path: string, propertyName: string): void {
    path = path.trim() === '/' ? '' : path;
    const requiredPath = sanitizeJsonPath(`${path}/required`);
    const requiredIndex = getValueAtPath(this.schema, requiredPath)?.indexOf(propertyName);
    if (requiredIndex > -1) {
      applyPatch(this.schema, [{ op: 'remove', path: `${requiredPath}/${requiredIndex}` }]);
    }
  }

  renameProperty(path: string, oldName: string, newName: string): void {
    const propertyPath = sanitizeJsonPath(`${path}/properties/${oldName}`);
    const newPath = sanitizeJsonPath(`${path}/properties/${newName}`);
    const propertyValue = getValueAtPath(this.schema, propertyPath);
    if (typeof propertyValue !== 'undefined') {
      applyPatch(this.schema, [{ op: 'remove', path: propertyPath }]);
      applyPatch(this.schema, [{ op: 'add', path: newPath, value: propertyValue }]);
      this.updateRequiredField(path, oldName, newName);
    }
  }

  rename(path: string, value: string): void {
    const newPath = sanitizeJsonPath(`${dirname(path)}/${value}`);
    const currentValue = getValueAtPath(this.schema, path);
    if (typeof currentValue !== 'undefined') {
      applyPatch(this.schema, [{ op: 'remove', path }]);
      applyPatch(this.schema, [{ op: 'add', path: newPath, value: currentValue }]);
    }
  }

  replace(path: string, value: string): void {
    const currentValue = getValueAtPath(this.schema, path);
    if (typeof currentValue !== 'undefined') {
      applyPatch(this.schema, [{ op: 'replace', path, value }]);
    }
  }

  add(path: string, value: string): void {
    applyPatch(this.schema, [{ op: 'add', path, value }]);
  }

  remove(path: string): void {
    const currentValue = getValueAtPath(this.schema, path);
    if (typeof currentValue !== 'undefined') {
      applyPatch(this.schema, [{ op: 'remove', path }]);
    }
  }

  private updateRequiredField(path: string, oldName: string, newName: string): void {
    const requiredPath = sanitizeJsonPath(`${path}/required`);
    const requiredArray = getValueAtPath(this.schema, requiredPath);
    if (requiredArray && requiredArray.includes(oldName)) {
      const index = requiredArray.indexOf(oldName);
      if (index !== -1) {
        applyPatch(this.schema, [
          { op: 'replace', path: `${requiredPath}/${index}`, value: newName }
        ]);
      }
    }
  }

}

export default JSONSchemaPatch;
