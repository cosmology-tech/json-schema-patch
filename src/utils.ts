import { JSONSchemaPatchOperation } from "src";

interface JSONSchema {
  properties?: { [key: string]: JSONSchema };
  items?: JSONSchema | JSONSchema[];
  $defs?: { [key: string]: JSONSchema };
  $ref?: string;
  [key: string]: any;
}

export interface PropertyPath {
  type: '$def' | 'property' | '$ref';
  path: string;
  name: string;
}

export type TransformFunction = (str: string) => string;

function sortByDeepestPath(propertyPaths: PropertyPath[]): PropertyPath[] {
  return propertyPaths.sort((a, b) => {
    const depthA = a.path.split('/').length;
    const depthB = b.path.split('/').length;
    return depthB - depthA; // Descending order
  });
}

function sortOpsByDeepestPath(propertyPaths: JSONSchemaPatchOperation[]): JSONSchemaPatchOperation[] {
  return propertyPaths.sort((a, b) => {
    const depthA = a.path.split('/').length;
    const depthB = b.path.split('/').length;
    return depthB - depthA; // Descending order
  });
}

function findProps(schema: JSONSchema, currentPath: string = ''): PropertyPath[] {
  let paths: PropertyPath[] = [];
  const name = basename(currentPath);
  const parent = dirname(currentPath);

  if (!(name === '' && currentPath === '')) {
    const parentName = basename(parent);
    if (parentName == '$defs') {
      paths.push({ type: '$def', path: currentPath, name });
    } else if (schema.$ref) {
      paths.push({ type: '$ref', path: currentPath, name: schema.$ref });
      paths.push({ type: 'property', path: currentPath, name });
    } else {
      paths.push({ type: 'property', path: currentPath, name });
    }
  }

  // Process properties
  if (schema.properties) {
    for (const [key, value] of Object.entries(schema.properties)) {
      paths = paths.concat(findProps(value, `${currentPath}/properties/${key}`));
    }
  }

  // Process $defs
  if (schema.$defs) {
    for (const [key, value] of Object.entries(schema.$defs)) {
      paths = paths.concat(findProps(value, `${currentPath}/$defs/${key}`));
    }
  }

  // Process items in arrays or tuples
  if (schema.items) {
    if (Array.isArray(schema.items)) {
      schema.items.forEach((item, index) => {
        paths = paths.concat(findProps(item, `${currentPath}/items/${index}`));
      });
    } else {
      paths = paths.concat(findProps(schema.items, `${currentPath}/items`));
    }
  }

  return paths;
}

export function findAllProps(schema: JSONSchema, currentPath: string = ''): PropertyPath[] {
  return sortByDeepestPath(findProps(schema, currentPath));
}

export function createJSONSchemaPatchOperations(propertyPaths: PropertyPath[], transform: TransformFunction): JSONSchemaPatchOperation[] {
  const ops = propertyPaths.map(propertyPath => {
    switch (propertyPath.type) {
      case 'property':
        return createRenamePropertyOperation(propertyPath, transform);
      case '$def':
        return createReplaceDefOperation(propertyPath, transform);
      case '$ref':
        return createReplaceRefOperation(propertyPath, transform);
      default:
        throw new Error(`Unknown property type: ${propertyPath.type}`);
    }
  }).filter(Boolean);
  const final = sortOpsByDeepestPath(ops);
  const replaces = final.filter(o => o.op === 'replace');
  const rest = final.filter(o => o.op !== 'replace');
  return [
    ...rest,
    ...replaces
  ];

}

function createRenamePropertyOperation(propertyPath: PropertyPath, transform: TransformFunction): JSONSchemaPatchOperation {
  const grandParent = dirname(dirname(propertyPath.path));
  const oldName = basename(propertyPath.path);
  const newName = transform(oldName);

  if (!isSimpleKey(oldName) || oldName === newName) return;

  return {
    op: 'renameProperty',
    path: grandParent,
    value: { oldName, newName }
  };
}

function createReplaceRefOperation(propertyPath: PropertyPath, transform: TransformFunction): JSONSchemaPatchOperation {
  const newRef = transformJsonPath(propertyPath.name, transform);
  if (newRef === propertyPath.name) return;
  return {
    op: 'replace',
    path: transformJsonPath(propertyPath.path + '/$ref', transform),
    value: newRef
  };
}

function createReplaceDefOperation(propertyPath: PropertyPath, transform: TransformFunction): JSONSchemaPatchOperation {
  const oldName = basename(propertyPath.path);
  const newName = transform(oldName);
  if (!isSimpleKey(oldName) || oldName === newName) return;
  return {
    op: 'rename',
    path: propertyPath.path,
    value: transformJsonPath(propertyPath.name, transform)
  };
}

function transformJsonPath(path: string, transformFunc: (str: string) => string): string {
  const specialKeywords = ['#', '$defs', '$ref', 'definitions', 'properties'];
  return path.split('/').map(segment => {
    const normalizedSegment = segment.replace(/^#/, '');
    if (specialKeywords.includes(normalizedSegment) || !isSimpleKey(segment)) {
      return segment;
    }
    return transformFunc(segment);
  }).join('/');
}

// MARKED AS NOT DRY
// from strfy-js
export function isSimpleKey(key: string): boolean {
  return /^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(key);
}

export function dirname(path: string): string {
  return path.replace(/\/[^\/]*$/, ''); // Removes last segment after the last '/'
}

export function basename(path: string): string {
  // Matches the last segment after the final '/'
  return path.split('/').pop() || '';
}
// end from strfy-js