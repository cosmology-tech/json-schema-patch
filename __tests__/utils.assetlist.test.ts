
import assetList from '../__fixtures__/assetlist.schema.json';
import { createJSONSchemaPatchOperations, findAllProps } from '../src/utils';

function camelCaseTransform(key: string): string {
  return key.replace(/[-_\s]+(.)?/g, (_, c) => c ? c.toUpperCase() : '');
}

it('findAllProps', () => {
  expect(findAllProps(assetList)).toMatchSnapshot();
})


it('createJSONSchemaPatchOperations', () => {
  expect(createJSONSchemaPatchOperations(findAllProps(assetList), camelCaseTransform)).toMatchSnapshot();
})

