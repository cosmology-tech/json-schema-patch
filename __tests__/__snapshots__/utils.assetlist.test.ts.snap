// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createJSONSchemaPatchOperations 1`] = `
[
  {
    "op": "renameProperty",
    "path": "/$defs/asset/properties/images/items/properties/theme",
    "value": {
      "newName": "primaryColorHex",
      "oldName": "primary_color_hex",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/asset/properties/images/items/properties/theme",
    "value": {
      "newName": "darkMode",
      "oldName": "dark_mode",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/asset/properties/ibc",
    "value": {
      "newName": "sourceChannel",
      "oldName": "source_channel",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/asset/properties/ibc",
    "value": {
      "newName": "dstChannel",
      "oldName": "dst_channel",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/asset/properties/ibc",
    "value": {
      "newName": "sourceDenom",
      "oldName": "source_denom",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/ibc_transition/properties/counterparty",
    "value": {
      "newName": "chainName",
      "oldName": "chain_name",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/ibc_transition/properties/counterparty",
    "value": {
      "newName": "baseDenom",
      "oldName": "base_denom",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/ibc_transition/properties/counterparty",
    "value": {
      "newName": "channelId",
      "oldName": "channel_id",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/ibc_transition/properties/chain",
    "value": {
      "newName": "channelId",
      "oldName": "channel_id",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/ibc_cw20_transition/properties/counterparty",
    "value": {
      "newName": "chainName",
      "oldName": "chain_name",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/ibc_cw20_transition/properties/counterparty",
    "value": {
      "newName": "baseDenom",
      "oldName": "base_denom",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/ibc_cw20_transition/properties/counterparty",
    "value": {
      "newName": "channelId",
      "oldName": "channel_id",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/ibc_cw20_transition/properties/chain",
    "value": {
      "newName": "channelId",
      "oldName": "channel_id",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/non_ibc_transition/properties/counterparty",
    "value": {
      "newName": "chainName",
      "oldName": "chain_name",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/non_ibc_transition/properties/counterparty",
    "value": {
      "newName": "baseDenom",
      "oldName": "base_denom",
    },
  },
  {
    "op": "replace",
    "path": "/$defs/asset/properties/denom_units/items/$ref",
    "value": "#/$defs/denomUnit",
  },
  {
    "op": "renameProperty",
    "path": "/$defs/asset",
    "value": {
      "newName": "extendedDescription",
      "oldName": "extended_description",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/asset",
    "value": {
      "newName": "denomUnits",
      "oldName": "denom_units",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/asset",
    "value": {
      "newName": "typeAsset",
      "oldName": "type_asset",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/asset",
    "value": {
      "newName": "logoURIs",
      "oldName": "logo_URIs",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/asset",
    "value": {
      "newName": "coingeckoId",
      "oldName": "coingecko_id",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/pointer",
    "value": {
      "newName": "chainName",
      "oldName": "chain_name",
    },
  },
  {
    "op": "renameProperty",
    "path": "/$defs/pointer",
    "value": {
      "newName": "baseDenom",
      "oldName": "base_denom",
    },
  },
  {
    "op": "renameProperty",
    "path": "",
    "value": {
      "newName": "chainName",
      "oldName": "chain_name",
    },
  },
  {
    "op": "rename",
    "path": "/$defs/denom_unit",
    "value": "denomUnit",
  },
  {
    "op": "rename",
    "path": "/$defs/ibc_transition",
    "value": "ibcTransition",
  },
  {
    "op": "rename",
    "path": "/$defs/ibc_cw20_transition",
    "value": "ibcCw20Transition",
  },
  {
    "op": "rename",
    "path": "/$defs/non_ibc_transition",
    "value": "nonIbcTransition",
  },
]
`;

exports[`findAllProps 1`] = `
[
  {
    "name": "primary_color_hex",
    "path": "/$defs/asset/properties/images/items/properties/theme/properties/primary_color_hex",
    "type": "property",
  },
  {
    "name": "circle",
    "path": "/$defs/asset/properties/images/items/properties/theme/properties/circle",
    "type": "property",
  },
  {
    "name": "dark_mode",
    "path": "/$defs/asset/properties/images/items/properties/theme/properties/dark_mode",
    "type": "property",
  },
  {
    "name": "#/$defs/pointer",
    "path": "/$defs/asset/properties/images/items/properties/image_sync",
    "type": "$ref",
  },
  {
    "name": "png",
    "path": "/$defs/asset/properties/images/items/properties/png",
    "type": "property",
  },
  {
    "name": "svg",
    "path": "/$defs/asset/properties/images/items/properties/svg",
    "type": "property",
  },
  {
    "name": "theme",
    "path": "/$defs/asset/properties/images/items/properties/theme",
    "type": "property",
  },
  {
    "name": "source_channel",
    "path": "/$defs/asset/properties/ibc/properties/source_channel",
    "type": "property",
  },
  {
    "name": "dst_channel",
    "path": "/$defs/asset/properties/ibc/properties/dst_channel",
    "type": "property",
  },
  {
    "name": "source_denom",
    "path": "/$defs/asset/properties/ibc/properties/source_denom",
    "type": "property",
  },
  {
    "name": "png",
    "path": "/$defs/asset/properties/logo_URIs/properties/png",
    "type": "property",
  },
  {
    "name": "svg",
    "path": "/$defs/asset/properties/logo_URIs/properties/svg",
    "type": "property",
  },
  {
    "name": "website",
    "path": "/$defs/asset/properties/socials/properties/website",
    "type": "property",
  },
  {
    "name": "twitter",
    "path": "/$defs/asset/properties/socials/properties/twitter",
    "type": "property",
  },
  {
    "name": "chain_name",
    "path": "/$defs/ibc_transition/properties/counterparty/properties/chain_name",
    "type": "property",
  },
  {
    "name": "base_denom",
    "path": "/$defs/ibc_transition/properties/counterparty/properties/base_denom",
    "type": "property",
  },
  {
    "name": "channel_id",
    "path": "/$defs/ibc_transition/properties/counterparty/properties/channel_id",
    "type": "property",
  },
  {
    "name": "channel_id",
    "path": "/$defs/ibc_transition/properties/chain/properties/channel_id",
    "type": "property",
  },
  {
    "name": "path",
    "path": "/$defs/ibc_transition/properties/chain/properties/path",
    "type": "property",
  },
  {
    "name": "chain_name",
    "path": "/$defs/ibc_cw20_transition/properties/counterparty/properties/chain_name",
    "type": "property",
  },
  {
    "name": "base_denom",
    "path": "/$defs/ibc_cw20_transition/properties/counterparty/properties/base_denom",
    "type": "property",
  },
  {
    "name": "port",
    "path": "/$defs/ibc_cw20_transition/properties/counterparty/properties/port",
    "type": "property",
  },
  {
    "name": "channel_id",
    "path": "/$defs/ibc_cw20_transition/properties/counterparty/properties/channel_id",
    "type": "property",
  },
  {
    "name": "port",
    "path": "/$defs/ibc_cw20_transition/properties/chain/properties/port",
    "type": "property",
  },
  {
    "name": "channel_id",
    "path": "/$defs/ibc_cw20_transition/properties/chain/properties/channel_id",
    "type": "property",
  },
  {
    "name": "path",
    "path": "/$defs/ibc_cw20_transition/properties/chain/properties/path",
    "type": "property",
  },
  {
    "name": "chain_name",
    "path": "/$defs/non_ibc_transition/properties/counterparty/properties/chain_name",
    "type": "property",
  },
  {
    "name": "base_denom",
    "path": "/$defs/non_ibc_transition/properties/counterparty/properties/base_denom",
    "type": "property",
  },
  {
    "name": "contract",
    "path": "/$defs/non_ibc_transition/properties/counterparty/properties/contract",
    "type": "property",
  },
  {
    "name": "contract",
    "path": "/$defs/non_ibc_transition/properties/chain/properties/contract",
    "type": "property",
  },
  {
    "name": "#/$defs/denom_unit",
    "path": "/$defs/asset/properties/denom_units/items",
    "type": "$ref",
  },
  {
    "name": "items",
    "path": "/$defs/asset/properties/traces/items",
    "type": "property",
  },
  {
    "name": "items",
    "path": "/$defs/asset/properties/images/items",
    "type": "property",
  },
  {
    "name": "items",
    "path": "/$defs/asset/properties/keywords/items",
    "type": "property",
  },
  {
    "name": "items",
    "path": "/$defs/denom_unit/properties/aliases/items",
    "type": "property",
  },
  {
    "name": "deprecated",
    "path": "/$defs/asset/properties/deprecated",
    "type": "property",
  },
  {
    "name": "description",
    "path": "/$defs/asset/properties/description",
    "type": "property",
  },
  {
    "name": "extended_description",
    "path": "/$defs/asset/properties/extended_description",
    "type": "property",
  },
  {
    "name": "denom_units",
    "path": "/$defs/asset/properties/denom_units",
    "type": "property",
  },
  {
    "name": "type_asset",
    "path": "/$defs/asset/properties/type_asset",
    "type": "property",
  },
  {
    "name": "address",
    "path": "/$defs/asset/properties/address",
    "type": "property",
  },
  {
    "name": "base",
    "path": "/$defs/asset/properties/base",
    "type": "property",
  },
  {
    "name": "name",
    "path": "/$defs/asset/properties/name",
    "type": "property",
  },
  {
    "name": "display",
    "path": "/$defs/asset/properties/display",
    "type": "property",
  },
  {
    "name": "symbol",
    "path": "/$defs/asset/properties/symbol",
    "type": "property",
  },
  {
    "name": "traces",
    "path": "/$defs/asset/properties/traces",
    "type": "property",
  },
  {
    "name": "ibc",
    "path": "/$defs/asset/properties/ibc",
    "type": "property",
  },
  {
    "name": "logo_URIs",
    "path": "/$defs/asset/properties/logo_URIs",
    "type": "property",
  },
  {
    "name": "images",
    "path": "/$defs/asset/properties/images",
    "type": "property",
  },
  {
    "name": "coingecko_id",
    "path": "/$defs/asset/properties/coingecko_id",
    "type": "property",
  },
  {
    "name": "keywords",
    "path": "/$defs/asset/properties/keywords",
    "type": "property",
  },
  {
    "name": "socials",
    "path": "/$defs/asset/properties/socials",
    "type": "property",
  },
  {
    "name": "denom",
    "path": "/$defs/denom_unit/properties/denom",
    "type": "property",
  },
  {
    "name": "exponent",
    "path": "/$defs/denom_unit/properties/exponent",
    "type": "property",
  },
  {
    "name": "aliases",
    "path": "/$defs/denom_unit/properties/aliases",
    "type": "property",
  },
  {
    "name": "chain_name",
    "path": "/$defs/pointer/properties/chain_name",
    "type": "property",
  },
  {
    "name": "base_denom",
    "path": "/$defs/pointer/properties/base_denom",
    "type": "property",
  },
  {
    "name": "type",
    "path": "/$defs/ibc_transition/properties/type",
    "type": "property",
  },
  {
    "name": "counterparty",
    "path": "/$defs/ibc_transition/properties/counterparty",
    "type": "property",
  },
  {
    "name": "chain",
    "path": "/$defs/ibc_transition/properties/chain",
    "type": "property",
  },
  {
    "name": "type",
    "path": "/$defs/ibc_cw20_transition/properties/type",
    "type": "property",
  },
  {
    "name": "counterparty",
    "path": "/$defs/ibc_cw20_transition/properties/counterparty",
    "type": "property",
  },
  {
    "name": "chain",
    "path": "/$defs/ibc_cw20_transition/properties/chain",
    "type": "property",
  },
  {
    "name": "type",
    "path": "/$defs/non_ibc_transition/properties/type",
    "type": "property",
  },
  {
    "name": "counterparty",
    "path": "/$defs/non_ibc_transition/properties/counterparty",
    "type": "property",
  },
  {
    "name": "chain",
    "path": "/$defs/non_ibc_transition/properties/chain",
    "type": "property",
  },
  {
    "name": "provider",
    "path": "/$defs/non_ibc_transition/properties/provider",
    "type": "property",
  },
  {
    "name": "#/$defs/asset",
    "path": "/properties/assets/items",
    "type": "$ref",
  },
  {
    "name": "$schema",
    "path": "/properties/$schema",
    "type": "property",
  },
  {
    "name": "chain_name",
    "path": "/properties/chain_name",
    "type": "property",
  },
  {
    "name": "assets",
    "path": "/properties/assets",
    "type": "property",
  },
  {
    "name": "asset",
    "path": "/$defs/asset",
    "type": "$def",
  },
  {
    "name": "denom_unit",
    "path": "/$defs/denom_unit",
    "type": "$def",
  },
  {
    "name": "pointer",
    "path": "/$defs/pointer",
    "type": "$def",
  },
  {
    "name": "ibc_transition",
    "path": "/$defs/ibc_transition",
    "type": "$def",
  },
  {
    "name": "ibc_cw20_transition",
    "path": "/$defs/ibc_cw20_transition",
    "type": "$def",
  },
  {
    "name": "non_ibc_transition",
    "path": "/$defs/non_ibc_transition",
    "type": "$def",
  },
]
`;
